moshi/src/main/java/com/squareup/moshi/JsonAdapter.kt

```diff
@@ -0,0 +1,298 @@
+    }
+    return result
+  }
+
+  /** Encodes the given [value] with the given [writer]. */
+  @Throws(IOException::class)
+  public abstract fun toJson(writer: JsonWriter, value: T?)
+
+  @Throws(IOException::class)
+  public fun toJson(sink: BufferedSink, value: T?) {
+    val writer = JsonWriter.of(sink)
+    toJson(writer, value)
+  }
+
+  /** Encodes the given [value] into a String and returns it. */
+  @CheckReturnValue
+  public fun toJson(value: T?): String {
+    val buffer = Buffer()
+    try {
+      toJson(buffer, value)
+    } catch (e: IOException) {
+      throw AssertionError(e) // No I/O writing to a Buffer.
+    }
+    return buffer.readUtf8()
+  }
+
+  /**
+   * Encodes [value] as a Java value object comprised of maps, lists, strings, numbers,
+   * booleans, and nulls.
+   *
+   * Values encoded using `value(double)` or `value(long)` are modeled with the
+   * corresponding boxed type. Values encoded using `value(Number)` are modeled as a [Long] for boxed integer types
+   * ([Byte], [Short], [Integer], and [Long]), as a [Double] for boxed floating point types ([Float] and [Double]),
+   * and as a [java.math.BigDecimal] for all other types.
+   */
+  @CheckReturnValue
+  public fun toJsonValue(value: T?): Any? {
+    val writer = JsonValueWriter()
+    return try {
+      toJson(writer, value)
+      writer.root()
+    } catch (e: IOException) {
+      throw AssertionError(e) // No I/O writing to an object.
+    }
+  }
+
+  /**
+   * Decodes a Java value object from [value], which must be comprised of maps, lists,
+   * strings, numbers, booleans and nulls.
+   */
+  @CheckReturnValue
+  public fun fromJsonValue(value: Any): T? {
+    val reader = JsonValueReader(value)
+    return try {
+      fromJson(reader)
+    } catch (e: IOException) {
+      throw AssertionError(e) // No I/O reading from an object.
+    }
+  }
+
+  /**
+   * Returns a JSON adapter equal to this JSON adapter, but that serializes nulls when encoding
+   * JSON.
+   */
+  @CheckReturnValue
+  public fun serializeNulls(): JsonAdapter<T> {
+    val delegate: JsonAdapter<T> = this
+    return object : JsonAdapter<T>() {
+      override fun fromJson(reader: JsonReader) = delegate.fromJson(reader)
+
+      override fun toJson(writer: JsonWriter, value: T?) {
+        val serializeNulls = writer.getSerializeNulls()
+        writer.setSerializeNulls(true)
+        try {
+          delegate.toJson(writer, value)
+        } finally {
+          writer.setSerializeNulls(serializeNulls)
+        }
+      }
+
+      override val isLenient: Boolean
+        get() = delegate.isLenient
+
+      override fun toString() = "$delegate.serializeNulls()"
+    }
+  }
+
+  /**
+   * Returns a JSON adapter equal to this JSON adapter, but with support for reading and writing
+   * nulls.
+   */
+  @CheckReturnValue
+  public fun nullSafe(): JsonAdapter<T> {
+    return if (this is NullSafeJsonAdapter<*>) {
+      this
+    } else {
+      NullSafeJsonAdapter(this)
+    }
```

Comment:
nit -- curious about `when` vs `if` here --

```suggestion
    return when(this)
      is NullSafeJsonAdapter<*> -> this
      else -> NullSafeJsonAdapter(this)
```